### PR TITLE
`OfferingsManager`/`OfferingsFactory`: using `OfferingsResponse`

### PR DESCRIPTION
**Pull request by NachoSoto**
Depends on #1437.
This takes full advantage of `Decodable` when decoding and creating `Offerings`.

### CODE REVIEW

**Review by taquitos — APPROVED** [review]
Awesome. 🥅 ⚽ 🐐 🎉

File: Sources/Networking/Operations/GetOfferingsOperation.swift [hunk 1]
```diff
@@ -46,7 +46,7 @@ private extension GetOfferingsOperation {
 
         let request = HTTPRequest(method: .get, path: .getOfferings(appUserID: appUserID))
 
-        httpClient.perform(request, authHeaders: self.authHeaders) { (response: HTTPResponse<[String: Any]>.Result) in
+        httpClient.perform(request, authHeaders: self.authHeaders) { (response: HTTPResponse<OfferingsResponse>.Result) in
```
Comment (NachoSoto) [inline Sources/Networking/Operations/GetOfferingsOperation.swift:49]: This is all it takes to get `HTTPClient` to deserialize the response 🤓

**Comment by NachoSoto** [comment]
This is finished now 🎉

File: Sources/Purchasing/OfferingsFactory.swift [hunk 2]
```diff
@@ -17,66 +17,66 @@ import StoreKit
 
 class OfferingsFactory {
```
Comment (NachoSoto) [inline Sources/Purchasing/OfferingsFactory.swift:18]: No more magic strings in this file 😄

File: Sources/Purchasing/OfferingsManager.swift [hunk 3]
```diff
@@ -37,7 +37,7 @@ class OfferingsManager {
         self.productsManager = productsManager
     }
 
-    func offerings(appUserID: String, completion: ((Offerings?, Error?) -> Void)?) {
+    func offerings(appUserID: String, completion: ((Result<Offerings, Error>) -> Void)?) {
```
Comment (NachoSoto) [inline Sources/Purchasing/OfferingsManager.swift:40]: Finished using `Result` around `OfferingsManager` too.

File: Sources/Purchasing/OfferingsManager.swift [hunk 4]
```diff
@@ -125,42 +131,26 @@ private extension OfferingsManager {
...
         Logger.appleError(Strings.offering.fetching_offerings_error(error: error.localizedDescription))
         deviceCache.clearOfferingsCacheTimestamp()
-        dispatchCompletionOnMainThreadIfPossible(completion,
-                                                 offerings: nil,
-                                                 error: error)
+        dispatchCompletionOnMainThreadIfPossible(completion, result: .failure(error))
     }
 
-    func extractProductIdentifiers(fromOfferingsData offeringsData: [String: Any]) -> Set<String> {
-        // Fixme: parse Data directly instead of converting from Data to Dictionary back to Data
```
Comment (NachoSoto) [inline Sources/Purchasing/OfferingsManager.swift:149]: Fixed

File: Tests/UnitTests/Networking/Backend/BackendGetOfferingsTests.swift [hunk 5]
```diff
@@ -70,27 +70,27 @@ class BackendGetOfferingsTests: BaseBackendTests {
...
         let packageB = packages[1]
 
         expect(offerings).to(haveCount(1))
-        expect(offeringA["identifier"] as? String) == "offering_a"
-        expect(offeringA["description"] as? String) == "This is the base offering"
-        expect(packageA["identifier"]) == "$rc_monthly"
-        expect(packageA["platform_product_identifier"]) == "monthly_freetrial"
-        expect(packageB["identifier"]) == "$rc_annual"
-        expect(packageB["platform_product_identifier"]) == "annual_freetrial"
-        expect(result?.value?["current_offering_id"] as? String) == "offering_a"
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/Backend/BackendGetOfferingsTests.swift:93]: No more magic strings

File: Tests/UnitTests/Purchasing/OfferingsTests.swift [hunk 6]
```diff
@@ -202,11 +199,15 @@ class OfferingsTests: XCTestCase {
     }
 
     func testOfferingsIsNilIfNoOfferingCanBeCreated() throws {
```
Comment (NachoSoto) [inline Tests/UnitTests/Purchasing/OfferingsTests.swift:201]: This is kind of an end-to-end test including deserialization too.

**Comment by NachoSoto** [comment]
This is still waiting for review 🤓

File: Tests/UnitTests/Networking/Backend/BackendGetOfferingsTests.swift (hunk 5 above)
Comment (aboedo) [reply Tests/UnitTests/Networking/Backend/BackendGetOfferingsTests.swift:93]: 🎉🎉🎉

File: Sources/Networking/Responses/OfferingsResponse.swift [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        let identifier: String
+        let description: String
+        let packages: [Package]
+
+    }
+
+    let currentOfferingId: String?
+    let offerings: [Offering]
+
+}
```
Comment (aboedo) [inline Sources/Networking/Responses/OfferingsResponse.swift:38]: 😍

**Review by aboedo — APPROVED** [review]

**Pull request merged**